### DISCUSSION TITLE
파이썬 바인딩 교체

### DISCUSSION THREAD

**Discussion by hahnlee** [hahnlee/hwp-rs#41] [제안]
블로그에도 언급하였듯 파이썬을 비롯한 바인딩 관리가 가장 큰 어려움 입니다.

https://mozilla.github.io/uniffi-rs/

모질라의 uniffi가 관리포인트를 크게 줄여줄것 같아 교체를 검토하고 있습니다

**Comment by mrchypark** [comment]
R 바인딩 제작을 고려하고 있는 입장에서, 이 변경이 관련 작성에 영향을 미치는지 궁금합니다.
r 과 rust의 경우 https://github.com/extendr/extendr 를 사용하고 있습니다.
제가 uniffi-rs를 전혀 몰라서 그러는데, 혹시 제가 검토해야 하는 부분이 있을까요?
혹은 python 바인딩을 확장하는 선택이기 때문에 고려할 필요가 없는것일까요?
답변 미리 감사드립니다.

**Reply by hahnlee** [reply]
현재 python바인딩처럼 rust버전의 create를 재작성하며 사용할경우 (즉, Paragraph를 PyParagraph로 만드는 코드와 같을경우)에는 문제 없을것으로 보입니다.

다만,
- 이 과정이 일일히 이뤄져야하기에 코스트가 높고
- 현재 breaking change가 계속 발생함은 참고해주세요 (semver에 따라 0.x 에선 마이너 버전이 올라가도 BC가 있을 수 있습니다. 현재 main브랜치는 rust 기준으로 BC가 발생했습니다.)
  - https://github.com/hahnlee/hwp-rs/tree/v0.0.1 태그 릴리즈 버전으로 작업하는것이 좋을것 같습니다.
 
**추가로 아직 결정한건 아니며**, uniffi의 경우 js를 지원하지 않아 uniffi의 커스텀제너레이터 기능을 사용해 js 바인딩을 만드는걸 검토하고 있습니다. 아마 r도 비슷한 방법으로 비용을 줄일 수 있을것 같습니다.

**Reply by hahnlee** [reply]
아, 추가로 hwp-rs는 어디까지나 로우레벨 API만을 제공하려고 합니다. 이 점은 파이썬 바인딩 구현에 있어 약간 후회하는 점인데요.
파이선을 예로들면, hwp-rs는 urllib3 같은 역할을하고, urllib3를 기반으로 request 같은 라이브러리가 나왔듯, 
페이지 쿼리, 더 편리한 API등은 다른 라이브러리로 구현을 권하고 있습니다. 따라서 uniffi로 생성된 바인딩도 마찬가지로 hwp의 구조에 가까운 낮은수준의 바인딩을 제공할 예정입니다.

따라서 편리한 쿼리등을 제공하려면 uniffi를 쓰는 바인딩이 있어도 추가로 관리해야할것 같습니다.

**Reply by mrchypark** [reply]
상세한 대답 감사드립니다.

semver 기준으로 메이저 버전이 0 인 경우 모든 버전업에 bc가 있다고 저는 가정하고 있습니다. ㅎㅎ 빨리 메이저 버전이 올라가면 좋겠네요.
말씀하신 것 처럼 uniffi가 R을 지원하는 설계가 가능하다면 너무 좋겠습니다. 관련해서도 제가 찾아보면 도움이 되겠네요.

추가하신 내용도 충분히 이해하였습니다.

hwp-rs를 제작해주셔서 정말 감사드립니다.

**Reply by hahnlee** [reply]
extendr을 보니 더더욱 러스트로 작성하길 잘했다는 생각이 드네요 감사합니다!

**Comment by hahnlee** [comment]
우선 다음 버전까지는 진행하지 않도록 하겠습니다.
